mesonbuild/interpreterbase.py

```diff
@@ -1026,6 +1026,28 @@ def string_method_call(self, obj: str, method_name: str, posargs: T.List[TYPE_nv
             if not isinstance(cmpr, str):
                 raise InterpreterException('Version_compare() argument must be a string.')
             return mesonlib.version_compare(obj, cmpr)
+        elif method_name == 'substring':
+            if len(posargs) > 2:
+                raise InterpreterException('substring() takes maximum two arguments.')
+            start = 0
+            end = len(obj)
+            if len (posargs) > 0:
+              start = int(posargs[0])
+            if len (posargs) > 1:
+                end = int(posargs[1])
+            if start < 0:
+                start = 0
+            if end < 0:
+                if abs(end) <= len(obj):
+                    end = len(obj) - abs(end)
+                else:
+                    end = 0
```

Comment:
you don't need to implement it yourself, python will do that logic for you already. tbh you don't need to do any checks on start/end values, python does the right thing for any value ;-)